### ISSUE TITLE
Route update removes interfaces from OVS bridge

### ISSUE THREAD

**Issue by nmathew** [nmstate/nmstate#1947]
Nmstate version: 1.2.0
NetworkManager (version 1.34.0)
SLES15-sp3

Problem:
**When applying an update to route configuration, ovs interface gets deleted.**

Full configuration and log here: [conf gist](https://gist.github.com/nmathew/f93c72ed976490698c3f13850fc8b456)

Node boot configuration:
routes:
    config:
        - table-id: 254
          destination: 0.0.0.0/0
          next-hop-interface: bm25c_ext_iface
          next-hop-address: 10.221.150.97
        - table-id: 200
          destination: 0.0.0.0/0
          next-hop-interface: traffic5_iface
          next-hop-address: 16.16.5.129
        - table-id: 254
          destination: 16.16.5.0/24
          next-hop-interface: traffic5_iface
          next-hop-address: 16.16.5.129
route-rules:
    config:
        - ip-from: 5.4.3.0/24
          route-table: 200
        - ip-from: 5.4.3.0/24
          ip-to: 192.168.0.0/24
          priority: 29994
          route-table: 254

Then updated the routing configuration with below config

#cat nmstate-cfg.yaml
routes:
    config:
        - table-id: 254
          destination: 16.16.1.1/24
          next-hop-interface: traffic1_iface
          next-hop-address: 10.221.150.97
        - table-id: 255
          destination: 16.16.1.2/24
          next-hop-interface: traffic5_iface
          next-hop-address: 10.221.150.98
        - table-id: 256
          destination: 17.17.2.5/24
          next-hop-interface: nex_mgmt_iface
          next-hop-address: 10.221.150.99
        - table-id: 257
          destination: 17.17.1.4/24
          next-hop-interface: nex_iface
          next-hop-address: 10.221.150.100
using "nmstatectl set nmstate-cfg.yaml" which gave below errors

`
interface failed: reason=<enum NM_ACTIVE_CONNECTION_STATE_REASON_DEVICE_DISCONNECTED of type NM.ActiveConnectionStateReason><enum NM_DEVICE_STATE_REASON_DEPENDENCY_FAILED of type NM.DeviceStateReason>
`

Before applying routes:

#ovs-vsctl show
eee2c28e-aaf7-4969-985f-109e6429b337
    Bridge data_bridge
        Port nex_iface
            tag: 4036
            Interface nex_iface
                type: internal
        Port traffic1_iface
            tag: 4033
            Interface traffic1_iface
                type: internal
        Port bond0
            Interface bond0
                type: system
        Port traffic5_iface
            tag: 4039
            Interface traffic5_iface
                type: internal
        Port nex_mgmt_iface
            tag: 4037
            Interface nex_mgmt_iface
                type: internal
        Port bm25c_ext_iface
            tag: 4030
            Interface bm25c_ext_iface
                type: internal
    ovs_version: "2.14.2"

After Applying routes:

#ovs-vsctl show
eee2c28e-aaf7-4969-985f-109e6429b337
    Bridge data_bridge
        Port traffic5_iface
            tag: 4039
            Interface traffic5_iface
                type: internal
        Port bond0
            Interface bond0
                type: system
        Port nex_mgmt_iface
            tag: 4037
            Interface nex_mgmt_iface
                type: internal
        Port nex_iface
            tag: 4036
            Interface nex_iface
                type: internal
        Port bm25c_ext_iface
            tag: 4030
            Interface bm25c_ext_iface
                type: internal
    ovs_version: "2.14.2"

traffic1_iface is missing.

ovs-vswitchd.log:
2022-06-24T09:50:46.214Z|00066|bridge|INFO|bridge data_bridge: deleted interface nex_iface on port 4
2022-06-24T09:50:46.307Z|00067|bridge|INFO|bridge data_bridge: deleted interface traffic5_iface on port 5
2022-06-24T09:50:46.307Z|00068|bridge|INFO|bridge data_bridge: deleted interface traffic1_iface on port 2
2022-06-24T09:50:46.307Z|00069|bridge|INFO|bridge data_bridge: deleted interface nex_mgmt_iface on port 3
2022-06-24T09:50:46.573Z|00070|bridge|INFO|bridge data_bridge: added interface nex_iface on port 7
2022-06-24T09:50:46.610Z|00071|bridge|INFO|bridge data_bridge: added interface traffic5_iface on port 8
2022-06-24T09:50:46.619Z|00072|bridge|INFO|bridge data_bridge: added interface nex_mgmt_iface on port 9

it is not creating the traffic1_iface back.

NetworkManager reports:
"device (traffic1_iface): state change: prepare -> failed (reason 'dependency-failed', sys-iface-state: 'managed')"

2022-06-24T09:50:46.465770+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.4655] device (traffic1_iface): Activation: starting connection 'traffic1_iface-if' (b38c394a-50b4-45f0-86c5-d36963d6d64f)
2022-06-24T09:50:46.467245+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.4671] device (traffic5_iface): state change: disconnected -> prepare (reason 'none', sys-iface-state: 'managed')
2022-06-24T09:50:46.467941+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.4678] device (traffic5_iface): state change: prepare -> config (reason 'none', sys-iface-state: 'managed')
2022-06-24T09:50:46.495487+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.4953] device (traffic1_iface): state change: disconnected -> prepare (reason 'none', sys-iface-state: 'managed')
2022-06-24T09:50:46.495851+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.4957] device (traffic1_iface): state change: prepare -> failed (reason 'dependency-failed', sys-iface-state: 'managed')
2022-06-24T09:50:46.496316+00:00 pool2-bm25c6 NetworkManager[3870]: <warn>  [1656064246.4961] device (traffic1_iface): Activation: failed for connection 'traffic1_iface-if'
2022-06-24T09:50:46.496656+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.4962] device (nex_mgmt_iface): state change: deactivating -> disconnected (reason 'new-activation', sys-iface-state: 'managed')
2022-06-24T09:50:46.533569+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.5333] device (nex_mgmt_iface): Activation: starting connection 'nex_mgmt_iface-if' (d4cc9a95-51bc-4918-b061-325a3204469d)
2022-06-24T09:50:46.534557+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.5344] device (traffic5_iface): state change: config -> ip-config (reason 'none', sys-iface-state: 'managed')
2022-06-24T09:50:46.535606+00:00 pool2-bm25c6 NetworkManager[3870]: <info>  [1656064246.5355] device (traffic5_iface): Activation: connection 'traffic5_iface-if' enslaved, continuing activation

From syslog:
2022-06-24T09:50:46.211523+00:00 pool2-bm25c6 dbus-daemon[2349]: [system] Activating via systemd: service name='org.freedesktop.nm_dispatcher' unit='dbus-org.freedesktop.nm-dispatcher.service' requested by ':1.21' (uid=0 pid=3870 comm="/usr/sbin/NetworkManager --no-daemon ")
2022-06-24T09:50:46.216038+00:00 pool2-bm25c6 kernel: [86364.266749] device nex_iface left promiscuous mode
2022-06-24T09:50:46.217556+00:00 pool2-bm25c6 systemd[1]: Starting Network Manager Script Dispatcher Service...
2022-06-24T09:50:46.229639+00:00 pool2-bm25c6 dbus-daemon[2349]: [system] Successfully activated service 'org.freedesktop.nm_dispatcher'
2022-06-24T09:50:46.230265+00:00 pool2-bm25c6 systemd[1]: Started Network Manager Script Dispatcher Service.
2022-06-24T09:50:46.308039+00:00 pool2-bm25c6 kernel: [86364.359717] device traffic5_iface left promiscuous mode
2022-06-24T09:50:46.384034+00:00 pool2-bm25c6 kernel: [86364.432628] device traffic1_iface left promiscuous mode
2022-06-24T09:50:46.468035+00:00 pool2-bm25c6 kernel: [86364.516548] device nex_mgmt_iface left promiscuous mode
2022-06-24T09:50:46.575003+00:00 pool2-bm25c6 systemd-udevd[47700]: Using default interface naming scheme 'sle15-sp3'.
2022-06-24T09:50:46.576062+00:00 pool2-bm25c6 kernel: [86364.624915] device nex_iface entered promiscuous mode
2022-06-24T09:50:46.582031+00:00 pool2-bm25c6 systemd-udevd[47700]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
2022-06-24T09:50:46.582268+00:00 pool2-bm25c6 systemd-udevd[47700]: nex_iface: Could not generate persistent MAC: No data available
2022-06-24T09:50:46.611382+00:00 pool2-bm25c6 systemd-udevd[47700]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
2022-06-24T09:50:46.611617+00:00 pool2-bm25c6 systemd-udevd[47700]: traffic5_iface: Could not generate persistent MAC: No data available
2022-06-24T09:50:46.612048+00:00 pool2-bm25c6 kernel: [86364.662156] device traffic5_iface entered promiscuous mode
2022-06-24T09:50:46.620029+00:00 pool2-bm25c6 kernel: [86364.671037] device nex_mgmt_iface entered promiscuous mode
2022-06-24T09:50:46.620636+00:00 pool2-bm25c6 systemd-udevd[47701]: Using default interface naming scheme 'sle15-sp3'.
2022-06-24T09:50:46.620913+00:00 pool2-bm25c6 systemd-udevd[47701]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
2022-06-24T09:50:46.621075+00:00 pool2-bm25c6 systemd-udevd[47701]: nex_mgmt_iface: Could not generate persistent MAC: No data available

Any help?

**Comment by nmathew** [comment]
@cathay4t @ffmancera  any pointers?

**Comment by cathay4t** [comment]
Hi @nmathew This looks like a NetworkManager bug. Could you enable trace log of NetworkManager by placing below content to file `/etc/NetworkManager/conf.d/debug.conf`:

```
[logging]
level=TRACE
domains=ALL
```

NetworkManager daemon restart is required to enable trace log.

I saw similar issue been solved by NetworkManager in 1.36 branch. If possible, please try upgrade to NetworkManager 1.36.6.

Meanwhile, your second state yaml is invalid. In my VM(with NM 1.39), nmstate complains about this route:

```
  - destination: 16.16.1.0/24
    next-hop-address: 10.221.150.98
    next-hop-interface: traffic5_iface
    table-id: 255
```

This is because there is no direct route to `10.221.150.98`, so kernel rejected this route.

**Comment by cathay4t** [comment]
@bengal is there NM patch regarding OVS `NM_ACTIVE_CONNECTION_STATE_REASON_DEVICE_DISCONNECTED` not been backported to NM 1.34 branch(this branch has not been updated since Jan 2022)?

**Comment by bengal** [comment]
Hi, there are some ovs-related fixes in 1.36, but it's not clear that they solve this problem because they are mostly related to service shutdown and externally removed interfaces. I would be useful to confirm whether the problem is solved on 1.36, or to see debug logs.

**Comment by nmathew** [comment]
This is working properly with NM 1.36.6, nmstate 1.2.0 and nispor 1.2.7, Thanks @cathay4t .
Closing the issue
